## Examen de Xarxes  Gener 2022

### Segon Parcial

1.- Donada la IPv4 129.234.55.9, indica:
* A quina classe  hi pertany i quina serà la seva màscara inicial? Quina part de l’adreça correspon a l’usuari i quina identifica la xarxa a la que pertany? Quants usuaris més es poden tenir en la xarxa associada a l’adreça de xarxa? (1 punt)
* Volem agrupar els usuaris en subxarxes de 254 adreces. Quantes subxarxes podem tenir? Cal modificar la màscara d’alguna manera per poder fer aquest exercici? Raona les respostes (1punt)
* A partir de una de les subxarxes anteriors, feu un subnetting variable per encabir una xarxa amb 100 usuaris i 4 xarxes amb 8 usuaris. Es possible fer aquesta divisió? Queden IPs lliures? Raona les respostes (1 punt)


In [3]:
# Agafem la primera part de l'adreça IP

a = 129
print("transformem el valor decimal de ",a, " en binari => ", bin(a))

transformem el valor decimal de  129  en binari =>  0b10000001


In [4]:
# Donat que comença per 10 vol dir que és CLASSE B
# Màscara = 255.255.0.0
# Adreça de xarxa => 129.234.0.0
# Adreça de host => 55.9
print('Podem tenir ', 2**16 -2 ," usuaris")

Podem tenir  65534  usuaris


El nombre total d'usuaris serà 65534, com que tenim assignada la ip 55.9 ens queden per assignar 65533

Volem crear subxarxes de 254 adreces per usuaris => $2^8 -2 $ => Agafem 8 bits
Ens queden per tant $2^8$ adreces per subxarxes => 256

Les adreces seran:

* 129.234.0.0/24
* 129.234.1.0/24
* 129.234.2.0/24
* ...
* 129.234.255.0/24
La màscara és 255.255.255.0 => Es a dir, /24

Agafo una adreça qualsevol, la primera:

129.234.0.0/24

**Primera subxarxa:**
129.234.0.0/25 amb els usuaris 129.234.0.1 fins al 129.234.0.126 es a dir tinc 128-2 = 126 adreces per usuaris.
**Resta de subxarxes: 4 subxarxes de 8 usuaris**

necessitem 2 bits per les quatre subxarxes. Per tant podriem fer:

129.234.0."11xxYYYY/28 => volen dir les x les diferents subxarxes i les Y els usuaris, que requereixen 4 bits, ja que $2^4 - 2 = 14$. Per tant tindrem:
* **129.234.0.192/28**
* i els usuaris seran 129.234.0.193/28
* 129.234.0.194/28
* 129.234.0.206/28
* **129.234.0.208/28**
* i els usuaris per aquesta seran 129.234.0.209/28
* fin 129.234.0.222/28
* **129.234.0.224/28**
* i els usuaris van des de 129.234.0.225/28 fins a 129.234.0.228/28
* **129.234.0.240/28**
* i els usuaris van des de 129.224.0.241/28 fins a 129.234.0.254/28

Tal i com es pot veure queden adreces IP lliures




2.- L’adreça IP 2001:3C4D:0DB8:0015::1A2F:1A2B identifica la id del prefixe de lloc, la id de la subxarxa i la id de l’usuari final. Quantes subxarxes es poden crear amb una IPv6? Raona la resposta. En aquest cas particular, quina subxarxa estem fent servir? (1 punt)

`El camp associat a les subxarxes inicialment és el quart i corresponen a 16 bits. En aquest cas correspon al valor 0015`

`ID Prefixe del lloc: 2001:3c4D:0DB8 (48 bits)`
`ID Usuari: 0000:0000:1A2F:1A2B (64 bits)`

`D'entrada es poden crear, amb una adreça d'aquestes fins a $2^{16}$ adreces`

3.- L’expedició localitzada a l’Antartida de la UB que estudia el canvi climàtic ens contacta per muntar un enllaç “wireless” punt a punt, amb una velocitat de transferència de 500 Mb/s. La mida de la trama és de 1024 Bytes. La mida de la trama de ACK és de 128 Bytes, la distància entre els dos punts és de 15.500 km aproximadament. La velocitat de propagació és de 3E8 m/s. La mida de la finestra K = 16 i la probabilitat d’error és del 0.1%. L’efecte de la trama ACK en l’eficiència, és menyspreable? Raona la resposta. Si considerem un protocol de repetició selectiva, quina eficiència ens permet el sistema? Raona la resposta (3 punts)

In [4]:
C = 500E6     # bits per segon
L = 1024*8    # bits
Lack = 128*8  # bits
d = 7845E3   # metres de distància
v = 3E8       # metres per segon
k = 1128      # finestra
Pe = 0.1E-2   # Probabilitat d'error de bit

In [5]:
# Veiem si la a es menyspreable.
# Calculem la a sense considerar ACK

Ttx = L/C
Tprop = d/v
Tack = Lack/C
print('temps de tx %f, temps de prop %f i temps de ack %f'%(Ttx,Tprop,Tack))
a = Tprop/Ttx
print('a sense considerar ACK', a)

# Ara calculem la a consinderant ACK

b = (Tprop + Tack/2)/Ttx
print('a considerant ACK', b)

temps de tx 0.000016, temps de prop 0.026150 i temps de ack 0.000002
a sense considerar ACK 1596.0693359374998
a considerant ACK 1596.1318359374998


In [6]:
print ('k >= 1+2a? ', (k>= (1+2*a)))
print(1+2*a)

k >= 1+2a?  False
3193.1386718749995


In [7]:
U = (1-Pe)*k/(1+2*a)
print('El valor de la Eficiencia és: ', U)

El valor de la Eficiencia és:  0.35290418481521973


`el valor de l'eficiència és igual a 0.25%. Aquest valor tant baix es deu a la gran distància que hi ha entre l'Antartida i Barcelona. Per incrementar-la hauriem d'incrementar el valor de la finestra`

4.- S’envia la trama següent: 37h 4Ah 5Bh i afegim per detectar errors una generador G(x) = X8+X. Indica el format de la trama que finalment s’envia. Valida el resultat en recepció. (1 punt)

`11 0111 0100 1010 0101 1011 / 1 0000 0010`

`11 0111 0100 1010 0101 1011 0000 0000`  
`10 0000 010`  
`01 0111 0000`  
`x1 0000 0010`  
`x0 0111 0010 10`  
`xx x100 0000 10`  
`xx x011 0010 001`  
`xx xx10 0000 010`  
`xx xx01 0010 0110`  
`xx xxx1 0000 0010`  
`xx xxx0 0010 0100 010`  
`xx xxxx xx10 0000 010`  
`xx xxxx xx00 0100 0001 10`  
`xx xxxx xxxx x100 0000 10`  
`xx xxxx xxxx x000 0001 0011 0000`  
`xx xxxx xxxx xxxx xxx1 0000 0010`  
`xx xxxx xxxx xxxx xxx0 0011 0010 000`  
`xx xxxx xxxx xxxx xxxx xx10 0000 010`  
`xx xxxx xxxx xxxx xxxx xx01 0010 0100`  
`xx xxxx xxxx xxxx xxxx xxx1 0000 0010`  
`xx xxxx xxxx xxxx xxxx xxx0 0010 0110`  

`Enviem la següent trama: `
`11 0111 0100 1010 0101 1011 0010 0110`


__Exercici 3__. Que fa el protocol NAT? On es fa servir? Explica el tipus de NAT que heu vist a pràctiques

`El protocol NAT (Network Address Translation) s'encarrega de traduir adreces IP privades en adreces IP públiques i al revés. Es pot aplicar allà on tinguis un router amb capacitat de fer-ho, però normalment, la xarxa LAN opera amb adreces privades i és el router de sortida de la LAN qui s'encarrega de fer aquesta conversió. Actualment, donat que ja no hi ha IPs públiques, qui s'encarrega de fer aquesta conversió és el ISP`

Pel que fa a la part de NAT vista a pràctiques, entenc que ho podeu revisar vosaltres mateixos, oi? 

__Exercici 4__. Quina adreça IP es fa servir com localhost? Explica perquè es pot fer servir

`L'adreça IP que s'utilitza com localhost és la 127.0.0.1  
Es fa servir per analitzar si tenim algun problema en el protocol TCP/IP del nostre PC, executant la comanda:  
ping 127.0.0.1  
Si veiem que no tenim connexió, tindrem un problema. També serveix per exemple per muntar sockets entre programes que corren en el nostre PC i muntar per exemple un client servidor senzillet.`

__Exercici 6__. Que fa la capa d'enllaç?

`La capa d'enllaç s'encarrega de gestionar les comunicacions punt a punt. Això implica fer servir el mateix format de TRAMA, gestionar els ERRORS i el CONTROL DE FLUX. Sempre punt a punt.`